iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java

```diff
@@ -752,6 +753,39 @@ private void addCustomSortExpr(Table table,  org.apache.hadoop.hive.ql.metadata.
     ).collect(Collectors.toList()));
   }
 
+  @Override
+  public boolean supportsPartitionAwareOptimization(org.apache.hadoop.hive.ql.metadata.Table table) {
+    if (hasUndergonePartitionEvolution(table)) {
+      // Don't support complex cases yet
+      return false;
+    }
+    final List<TransformSpec> specs = getPartitionTransformSpec(table);
+    // Currently, we support the only bucket transform
+    return specs.stream().anyMatch(HiveIcebergStorageHandler::isBucket);
+  }
+
+  @Override
+  public PartitionAwareOptimizationCtx createPartitionAwareOptimizationContext(
+      org.apache.hadoop.hive.ql.metadata.Table table) {
+    // Currently, we support the only bucket transform
+    final List<String> bucketColumnNames = Lists.newArrayList();
+    final List<Integer> numBuckets = Lists.newArrayList();
+    getPartitionTransformSpec(table).stream().filter(HiveIcebergStorageHandler::isBucket).forEach(spec -> {
+      bucketColumnNames.add(spec.getColumnName());
+      numBuckets.add(spec.getTransformParam().get());
+    });
+
+    if (bucketColumnNames.isEmpty()) {
+      return null;
+    }
+    final IcebergBucketFunction bucketFunction = new IcebergBucketFunction(bucketColumnNames, numBuckets);
+    return new PartitionAwareOptimizationCtx(bucketFunction);
+  }
+
+  private static boolean isBucket(TransformSpec spec) {
```

Comment:
could we move this method under the TransformSpec.isBucket()?

Reply:
how about IcebergTableUtil then?

Reply:
TransformSpec was originally introduced for iceberg, see HIVE-25314

Reply:
I initially thought we could reuse TransformSpec to implement the Hudi storage format, for example. To clarify, I don't know even what spec Hudi supports, and I don't have a strong preference. I will move the logic

Reply:
Sorry, now I think we may keep this one here. HIVE-28411 supports only bucket transforms but I think we can finally support other transforms such as DAY or HOUR. We will no longer need `isBucket` someday.

Reply:
ok

Reply:
could we at least move it to `IcebergTableUtil`? but I still think it belongs to `TransformSpec` context

Reply:
Moved
https://github.com/apache/hive/pull/5409/commits/5c763b976733f106749637e5bb8c344dbf05bc7b